File: go/change/change.go
package change

import (
	"errors"
)

func Change(coins []int, target int) ([]int, error) {
	change := []int{}
	if target < 0 {
		return change, errors.New("target cannot be negative")
	} else if target == 0 {
		return change, nil
	}
	for i := len(coins) - 1; i >= 0; i-- {
		best, err := Change(coins[:i+1], target - coins[i])
		if err == nil && (len(best) < len(change)-1 || len(change) == 0) {
			change = append(best, coins[i])
		}
		if len(change) > 0 && i != 0 && target/coins[i-1] >= len(change) {
			break
		}
	}
	if len(change) == 0 {
		return change, errors.New("exact change cannot be given")
	}
	return change, nil
}

File: python/change/change.py
def find_fewest_coins(coins, target):
    if target < 0:
        raise ValueError("target can't be negative")
    if target == 0:
        return []
    change = []
    for i in range(len(coins) - 1, -1, -1):
        try:
            best = find_fewest_coins(coins[:i+1], target - coins[i])
            if len(best) < len(change)-1 or len(change) == 0:
                best.append(coins[i])
                change = best
        except:
            pass
        finally:
            if len(change) > 0 and i != 0 and target/coins[i-1] >= len(change):
                break
    if len(change) == 0:
        raise ValueError("can't make target with given coins")
    return change
